src/search.rs

```diff
@@ -0,0 +1,141 @@
+use crate::document::advanced::DocumentEncryptUnmanagedResult;
+use crate::document::advanced::*;
+use crate::document::DocumentEncryptOpts;
+use crate::internal::{take_lock, IronOxideErr};
+use crate::{GroupId, IronOxide};
+use async_trait::async_trait;
+use rand::{self, RngCore};
+use std::convert::{TryFrom, TryInto};
+use std::ops::DerefMut;
+
+use ironcore_search_helpers::generate_hashes_for_string;
+
+type Result<A> = std::result::Result<A, IronOxideErr>;
+///The required length of the salt.
+const REQUIRED_LEN: usize = 32;
+
+///The result of creating a new index as well as indexing on a particular phrase.
+///If you only want to create the index, see create_index.
+pub struct CreatedIndexResult {
+    pub encrypted_salt: DocumentEncryptUnmanagedResult,
+    pub sdk: IronSimpleSearch,
+}
+
+///Trait which gives the ability to create an index
+#[async_trait]
+pub trait SimpleSeachInitialize {
+    ///Given the encrypted salt and the edeks, decrypt them and give back the IronSimpleSearch object.
+    async fn initialize_search_index(
+        &self,
+        encrypted_data: &[u8],
+        encrypted_deks: &[u8],
+    ) -> Result<IronSimpleSearch>;
+    ///Create an index and encrypt it to the provided group_id.
+    ///If you need to index terms immediately, see `create_index_and_initialize` which will return
+    ///the IronSimpleSearch for reuse.
+    async fn create_index(&self, group_id: &GroupId) -> Result<DocumentEncryptUnmanagedResult>;
+    ///Create an index and also index the term.
+    async fn create_index_and_initialize(&self, group_id: &GroupId) -> Result<CreatedIndexResult>;
+}
+
+#[async_trait]
+impl SimpleSeachInitialize for IronOxide {
+    async fn initialize_search_index(
+        &self,
+        encrypted_salt: &[u8],
+        encrypted_salt_deks: &[u8],
+    ) -> Result<IronSimpleSearch> {
+        let decrypted_value = self
+            .document_decrypt_unmanaged(encrypted_salt, encrypted_salt_deks)
+            .await?;
+        decrypted_value.decrypted_data().try_into()
+    }
+    async fn create_index(&self, group_id: &GroupId) -> Result<DocumentEncryptUnmanagedResult> {
+        let CreatedIndexResult { encrypted_salt, .. } =
+            self.create_index_and_initialize(group_id).await?;
+        Ok(encrypted_salt)
+    }
+    async fn create_index_and_initialize(&self, group_id: &GroupId) -> Result<CreatedIndexResult> {
+        let salt = {
```

Comment:
What's the benefit of the lambda function over just inlining?

Reply:
This isn't a lambda, just limited scoping of the mutable variable.